sotodlib/site_pipeline/make_read_det_match.py

```diff
@@ -0,0 +1,169 @@
+            dest_dataset = f"{ts.stream_id}_{ts.name}_mapping_v{mapping['version']}"
+            
+            if dest_dataset in db.get_entries() and not args.override:
+                logger.debug(f"Dataset {dest_dataset} already exists")
+                continue
+            
+            ## Setup Mapping for Each Mapping Type
+            map_maker = MapMaker(
+                north_is_highband=array["north_is_highband"],
+                array_name = array["name"],
+                mapping_strategy = mapping["strategy"],
+                dark_bias_lines=array["dark_bias_lines"],
+                **mapping["params"]
+            )
+
+            array_map=None
+            logger.info(f"Making Map for {ts.path}")
+            try:
+                array_map = map_maker.make_map_smurf(tunefile=ts.path)
+            except:
+                logger.warning(f"Map Maker Failed on {ts.path}")
+                continue
+
+            readout_ids, bands, channels = zip(*[
+                    (ch.name, ch.band, ch.channel) for ch in ts.channels
+                    ])
+            bands = np.array(bands)
+            channels = np.array(channels)
+
+            rs_list = core.metadata.ResultSet(
+                keys=["dets:det_id", 
+                      "dets:readout_id"]
+            )
+
+            ## loop through detector IDs and find readout ID matches
+            for tune in array_map:
+                det_id = tune.detector_id
+                if det_id is None:
+                    ## resonators smurf found that array map doesn't think 
+                    ## should exist (artifacts?)
+                    continue
+                i_did = np.where(det_info.dets.vals == det_id)[0]
+                if len(i_did) != 1:
+                    logger.warning(f"Map returned detector_id, {det_id}, "
+                                "not in database, what happened?")
+                    continue
+                i_did = i_did[0]
+                i_rid = np.where(np.all([
+                        bands == tune.smurf_band,
+                        channels == tune.smurf_channel], axis=0))[0]
+                if len(i_rid) != 1:
+                    logger.debug(f"Detector {det_id} not found in Tuneset {ts.name}")
+                    ## If we want to track which detectors do not have readout_ids it
+                    ## would go here, but has trouble with dets:readout_id loading in 
+                    ## the current framework
+                    continue
+                i_rid = i_rid[0]
+                #ch_info.readout_id[i_did] = readout_ids[i_rid]
+                rs_list.append( {"dets:det_id": det_id,
+                                 "dets:readout_id": readout_ids[i_rid]})
+
+            for rid in readout_ids:
+                if rid in rs_list["dets:readout_id"]:
+                    continue
+                rs_list.append( {"dets:det_id": "NO_MATCH",
+                                 "dets:readout_id": rid})
```

Comment:
My initial reaction to this was "agh, this introduces a single value into multiple rows of a field that is sort-of-supposed-to-be an "id" in the sense of globally unique."  This could cause trouble downstream, if code assumes det_id is unique.  Then I started thinking about how other metadata products that are indexed by det_id (e.g. focal plane positions) would lead to all the NO_MATCH detectors to be dropped, when the axes are combined... then I realized that it might be really useful for downstream products to be able to provide a default value to use, and that would naturally be done by having a row with dets:det_id=NO_MATCH.

So... after much editing, I think this is a good idea but please include this in the FLP interface description for this product.  And I'd like to include it in the metadata system tests.

Reply:
Done.